formats/protobuf/commonTest/src/kotlinx/serialization/protobuf/ProtobufOneOfTest.kt

```diff
@@ -0,0 +1,659 @@
+         * 6:LEN {"bar"}
+         * 3:LEN {"foo"}
+         */
+        ProtoBuf.decodeFromHexString<OneOfDataNullable>("32036261721a03666f6f").let {
+            assertEquals(OneOfDataNullable(StringType("bar"), "foo"), it)
+        }
+        /**
+         * 5:VARINT 42
+         * 3:LEN {"foo"}
+         */
+        ProtoBuf.decodeFromHexString<OneOfDataNullable>("282a1a03666f6f").let {
+            assertEquals(OneOfDataNullable(IntType(42), "foo"), it)
+        }
+        /**
+         * 3:LEN {"foo"}
+         */
+        ProtoBuf.decodeFromHexString<OneOfDataNullable>("1a03666f6f").let {
+            assertEquals(OneOfDataNullable(null, "foo"), it)
+        }
+    }
+
+    @Test
+    fun testOneOfToFlatInt() {
+        val dataInt = OneOfData(IntType(42), "foo")
+        ProtoBuf.encodeToByteArray(dataInt).let {
+            ProtoBuf.decodeFromByteArray<DataNoOneOf>(it).let { data ->
+                assertEquals(
+                    DataNoOneOf(42, "", "foo"), data
+                )
+            }
+        }
+    }
+
+    @Test
+    fun testOneOfToFlatString() {
+        val dataString = OneOfData(StringType("bar"), "foo")
+        ProtoBuf.encodeToByteArray(dataString).let {
+            ProtoBuf.decodeFromByteArray<DataNoOneOf>(it).let { data ->
+                assertEquals(
+                    DataNoOneOf(0, "bar", "foo"), data
+                )
+            }
+        }
+    }
+
+    @Test
+    fun testFlatIntToOneOf() {
+        val dataInt = DataNoOneOf(42, "", "foo")
+        ProtoBuf.encodeToByteArray(dataInt).let {
+            ProtoBuf.decodeFromByteArray<OneOfData>(it).let { data ->
+                assertEquals(
+                    OneOfData(IntType(42), "foo"), data
+                )
+            }
+        }
+    }
+
+    @Test
+    fun testFlatStringToOneOf() {
+        val dataString = DataNoOneOf(0, "bar", "foo")
+        ProtoBuf.encodeToByteArray(dataString).let {
+            ProtoBuf.decodeFromByteArray<OneOfData>(it).let { data ->
+                assertEquals(
+                    OneOfData(StringType("bar"), "foo"), data
+                )
+            }
+        }
+    }
+
+    @Test
+    fun testEncodeNestedStruct() {
+        val data = OneOfData(NestedIntType(IntType(32)), "foo")
+        ProtoBuf.encodeToHexString(OneOfData.serializer(), data).also {
+            println(it)
+            /**
+             * 7:LEN {5:VARINT 32}
+             * 3:LEN {"foo"}
+             */
+            assertEquals("3a0228201a03666f6f", it)
+        }
+    }
+
+    @Test
+    fun testDecodeNestedStruct() {
+        val data = OneOfData(NestedIntType(IntType(32)), "foo")
+        /**
+         * 7:LEN {5:VARINT 32}
+         * 3:LEN {"foo"}
+         */
+        ProtoBuf.decodeFromHexString<OneOfData>("3a0228201a03666f6f").also {
+            println(it)
+            assertEquals(data, it)
+        }
+    }
+
+    @Serializable
+    data class FailType(@ProtoNumber(8) val i: Int, @ProtoNumber(9) val j: Int) : IType
```

Comment:
I see that only encoding of this `FailType` fails, while decoding of `IType` happens successfully. Shoudn't we check that the number of elements is 1 for decoding, too? What happens if someone passes an input with `8: VARINT 32 3: LEN "foo"` or `8:VARINT 32 9: VARINT 33 3: LEN "foo"`?

Reply:
Both encoding and decoding of type `FailType` will fail, but havent written a test of decoding `FailType` (added now).